### ISSUE TITLE
Exact and approximate simulation methods for stochastic models

### ISSUE THREAD

**Issue by ben18785** [pints-team/pints#889]
For models that we use to exemplify PINTS' ABC capabilities - in particular, stochastic reaction equations - there are sometimes exact (computationally expensive) and approximate (computationally cheap) ways to simulate from these series. For example, the Gillespie algorithm is an exact SSA whereas tau leaping is approximate. Some ABC methods, such as lazy ABC, use both cheap and expensive simulation methods to simulate and so, think it'd be good to include these.

To this end, I would propose making an abstract toy model class for stochastic reaction equations. @danielfridman98 has started this work by coding up the stochastic degradation algorithm in lightning fashion!

**Comment by MichaelClerx** [comment]
Hmmmm! I've seen some other methods that use fast&cheap versus slow&exact, but should we be investigating these? I certainly don't have any models where this is possible!

**Comment by ben18785** [comment]
Yep, think we should -- in typical ABC problems they often exist. They're a
class of ABC methods that are of growing importance.

**Comment by MichaelClerx** [comment]
We should maybe discuss this at the next PINTS meeting: don't want to be PINTSing for the sake of it

**Comment by ben18785** [comment]
Happy to discuss! Definitely worth reading up on this though if you have
time -- the lazy ABC paper and lots of Ruth Baker's papers are based on
this idea.

Cheers,

Ben

**Comment by MichaelClerx** [comment]
Just some added info on optimising/sampling approximate (cheaper) functions.

There are optimisation methods called [Sequential Model-based Global Optimization](https://github.com/scikit-optimize/scikit-optimize) methods that do this.
In the context of ODEs gradient matching is another technique that uses a different score function to speed things up #22

**Comment by MichaelClerx** [comment]
Multi-fidelity methods used in cardiac activation time simulations https://doi.org/10.1016/j.jcp.2019.03.026
